core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -269,25 +313,20 @@ class ZooKeeperClient(connectString: String,
   private def initialize(): Unit = {
     if (!zooKeeper.getState.isAlive) {
       info(s"Initializing a new session to $connectString.")
-      var now = System.currentTimeMillis()
-      val threshold = now + connectionTimeoutMs
-      while (now < threshold) {
+      // retry forever until ZooKeeper can be instantiated
+      while (true) {
         try {
           zooKeeper.close()
           zooKeeper = new ZooKeeper(connectString, sessionTimeoutMs, ZooKeeperClientWatcher)
-          waitUntilConnected(threshold - now, TimeUnit.MILLISECONDS)
           return
         } catch {
-          case _: Exception =>
-            now = System.currentTimeMillis()
-            if (now < threshold) {
-              Thread.sleep(1000)
-              now = System.currentTimeMillis()
-            }
+          case e: Exception =>
+            info("Error when recreating ZooKeeper", e)
+            Thread.sleep(1000)
         }
       }
       info(s"Timed out waiting for connection during session initialization while in state: ${zooKeeper.getState}")
-      stateChangeHandlers.foreach {case (name, handler) => handler.onReconnectionTimeout()}
+      stateChangeHandlers.values.foreach(_.onReconnectionTimeout())
```

Comment:
I guess this code will never run now?

Reply:
Good point. We can just remove the onReconnectionTimeout() callback.